meson.build

```diff
@@ -1,47 +1,129 @@
 
-sources = [
-        'src/main.c'
+if not meson.is_cross_build()
+        error('Cross-compiling is required. Please run meson with the \'--cross-file targets/family/my-target-family.txt\' \'--cross-file targets/my-target.txt\' flag')
+endif
+
+# ---------------------------------------
+# Target independent sources & flags
+# ---------------------------------------
+
+source_dirs = [
+	'src',
+	'external/tinyusb/src',
+	'external/tinyusb/src/class/hid',
+	'external/tinyusb/src/common',
+	'external/tinyusb/src/device'
+]
+
+include_dirs = [
+	'src/target',
+	'external/tinyusb/src/osal'
 ]
 
 c_flags = [
-        '-Os'
+	'-Os',
+	'-nostdlib',
+	'-nostartfiles',
+	'-ffunction-sections',
+	'-fdata-sections',
+	'-std=gnu99',
+	'-Wpointer-arith',
+	'-Wundef',
+	'-Werror'
 ]
 
 ld_flags = [
-        '--specs=nosys.specs',
-        '-nostartfiles',
-        '-Wl,--gc-sections',
-        '-T' + meson.get_cross_property('linker') + '.ld'
+	'-lm',
+	'-lc',
+	'-lgcc',
+	'-lnosys',
+	'-nostdlib',
+	'-nostartfiles',
+	'-fdata-sections',
+	'-ffunction-sections',
+	'--specs=nano.specs',
+	'--specs=nosys.specs',
+	'-Wl,--gc-sections',
+	'-T../build-targets/ld/' + meson.get_cross_property('linker') + '.ld'
 ]
 
-if not meson.is_cross_build()
-        error('Cross-compiling is required. Please run meson with the \'--cross-file targets/family/my-target-family.txt\' \'--cross-file targets/my-target.txt\' flag')
+# ---------------------------------------
+# Target dependent sources & flags
+# ---------------------------------------
+
+target_family = meson.get_cross_property('family')
+
+if target_family == 'stm32f1'
+	include_dirs += [
+		'external/cmsis-5/CMSIS/Core/Include',
+		'external/cmsis-device-stm32f1/Include'
+	]
+	source_dirs += [
+		'src/drv/mcu/stm32f1',
+		'external/tinyusb/src/portable/st/stm32_fsdev'
+	]
+else
+	error('This family is not supported.')
+endif
+
+target = meson.get_cross_property('target')
+
+if target == 'blue-pill'
+	c_flags += [
+		'-DSTM32F103xB',
+		'-DHSE_VALUE=8000000UL'
+	]
+	source_dirs += [
+		'src/target/stm32f1'
+	]
+elif target == 'rival310'
+	c_flags += [
+		'-DSTM32F103x6',
+		'-DHSE_VALUE=12000000UL'
+	]
+	source_dirs += [
+		'src/target/rival310'
+	]
+else
+	error('This target is not supported.')
 endif
 
-# libopencm3 special handling
+sensor = meson.get_cross_property('sensor')
+
+if sensor == 'truemove3'
+	source_dirs += [
+		'src/drv/sensor/truemove3'
+	]
+elif sensor == 'none'
+else
+	error('This sensor is not supported.')
+endif
 
-prefix = meson.current_source_dir() / '/usr' / meson.get_cross_property('prefix')
-include_dir = prefix / 'include'
-lib_dir = prefix / 'lib'
+sources = []
 
-compiler = meson.get_compiler('c', native: true)
-libopencm3 = compiler.find_library(meson.get_cross_property('libopencm3_cpu'), dirs: lib_dir)
+foreach dir : source_dirs
+sources += run_command('find', dir, '-maxdepth', '1', '-type', 'f', '-name', '*.c').stdout().strip().split('\n')
```

Comment:
Not sure if there is a right way of doing this, but this works.

Reply:
Our sources should always have an explicit declaration (there's a reason meson does not have native support for this). To compile the vendor libraries, I would accept something like this.

Reply:
why does meson require that? i don't really love to explicitly declare every source, i find this much cleaner, but I'll accept it if it is to be that way.